### PR TITLE
Make access control strategy configurable

### PR DESCRIPTION
**Pull request by seanstory**
## Part of https://github.com/elastic/connectors-python/issues/1204

See https://elastic.slack.com/archives/C01795T48LQ/p1688745239319069

Before this change, the access control sync for Sharepoint Online would do something like:
```
for each site collection
  for each site
    for each item in the UserInformationList
      for each person in a group item
        for each group that person belongs to
          add_to_acl_list(group)
```          

This 5 levels of nesting is only efficient if each collection is relatively small. 

This PR adds an alternate approach:

```
for each user in list_users
  for each group that user belongs to
    add_to_acl_list(group)
```

This is likely more efficient at larger scales, but may be less efficient for large enterprises where the connector is only intended to sync a small number of small/private sites.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes

### CODE REVIEW

File: connectors/sources/sharepoint_online.py [hunk 1]
```diff
@@ -1174,97 +1194,121 @@ async def get_access_control(self):
...
+                return person_access_control_doc
 
-                                update_already_seen(member_email, member_username)
+        if not self.configuration["fetch_users_by_site"]:
+            self._logger.info("Fetching all users")
+            async for user in self.client.users():
+                user_doc = await process_user(user)
+                if not user_doc:
+                    continue
+                yield user_doc
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:1236]:
```suggestion
                if user_doc:
                    yield user_doc
```

nit: super small nit

File: connectors/sources/sharepoint_online.py [hunk 2]
```diff
@@ -1174,97 +1194,121 @@ async def get_access_control(self):
...
+                self._logger.debug(
+                    f"Looking at site collection for location: {site_collection.get('dataLocationCode')}"
+                )
+                async for site in self.client.sites(
+                    site_collection["siteCollection"]["hostname"],
+                    self.configuration["site_collections"],
+                ):
+                    site_id = site.get("id")
+                    self._logger.debug(f"Looking at site: {site_id}")
+                    if _already_seen(site_id):
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:1249]: Nice, so a site can be part of multiple site collections I suppose?

File: connectors/sources/sharepoint_online.py [hunk 3]
```diff
@@ -1174,97 +1194,121 @@ async def get_access_control(self):
...
+                                        await self._user_access_control_doc(owner)
+                                    )
+                                    if owner_access_control_doc:
+                                        yield owner_access_control_doc
+
+                        elif is_person(user):
+                            user_doc = await process_user(user)
+                            if not user_doc:
+                                continue
+                            yield user_doc
```
Comment (timgrein) [inline connectors/sources/sharepoint_online.py:1306]:
```suggestion
                            if user_doc:
                                yield user_doc
```

nit: just to make it consistent with the other places

**Review by timgrein — DISMISSED** [review]
LGTM, only two nits. Really nice investigation and solution 👏  
I guess we should've a corresponding PR in Kibana for the new RFC?

**Comment by seanstory** [comment]
> I guess we should've a corresponding PR in Kibana for the new RFC?

Yep. I have two other tasks on https://github.com/elastic/connectors-python/issues/1204#issuecomment-1625942321 if this PR is going in. And given that I haven't seen any objections, I think it will.

**Review by timgrein — APPROVED** [review]

**Pull request merged**